exonum/src/runtime/mod.rs

```diff
@@ -404,9 +410,33 @@ pub trait Runtime: Send + fmt::Debug + 'static {
         &mut self,
         snapshot: &dyn Snapshot,
         spec: &InstanceSpec,
-        status: InstanceStatus,
+        status: &InstanceStatus,
     );
 
+    /// Gets the migration script to migrate the data of the service to the state usable
+    /// by a newer version of the artifact.
+    ///
+    /// # Invariants Ensured by the Caller
+    ///
+    /// - `new_artifact` is deployed in the runtime
+    /// - `old_service` exists and is stopped
+    ///
+    /// # Return Value
+    ///
+    /// - An error signals that the runtime does not know how to migrate the service
+    ///   to a newer version.
+    /// - `Ok(Some(_))` provides a script to execute against service data. After the script
+    ///   is executed, the service will have its version updated to `end_version` from the script.
+    ///   Note that this version does not need to correspond to the version of `new_artifact`,
+    ///   or to a version of an artifact deployed on the blockchain in general.
+    /// - `Ok(None)` means that the service does not require data migration. The service will
+    ///   have its version updated to the version of `new_artifact` immediately.
+    fn migrate(
+        &self,
+        new_artifact: &ArtifactId,
+        old_service: &InstanceSpec,
+    ) -> Result<Option<MigrationScript>, InitMigrationError>;
```

Comment:
Maybe it makes sense to make `MigrationScript` an enum with structure like:
```rust
pub enum MigrationScript {
  NoOp, // Meaning that migration is fast-forward and no data migration required.
  Script(MigrationScriptData), // Contents of the current `MigrationScript`.
}
```

If this list of provided methods will be the same as for current `MigrationScript` (for `NoOp` they won't do anything), this will allow us to get rid of some `unwrap`s, make function signatures easier, and, I believe, the code overall will become more clean.

What do you think?

Reply:
I've thought about it (or, alternatively, creating a no-op `MigrationScript` constructor). The problem with this approach is that fast-forward migrations are *intentionally* processed differently from normal migrations (there is no hash agreement, no rollbacks / commits, etc.). `MigrationScript` as enum is slightly cleaner, but semantically `MigrationScript::NoOp` does not have the same fields / methods as the "ordinary" variant; for example, it does not have a description or `end_version`.

Reply:
Indeed. Well, we surely can think about this moment later.